server/src/org/apache/cloudstack/affinity/AffinityGroupServiceImpl.java

```diff
@@ -432,12 +432,25 @@ public UserVm updateVMAffinityGroups(Long vmId, List<Long> affinityGroupIds) {
                 throw new InvalidParameterValueException("Unable to find affinity group by id " + affinityGroupId);
             } else {
                 // verify permissions
-                _accountMgr.checkAccess(caller, null, true, owner, ag);
-                // Root admin has access to both VM and AG by default, but make sure the
-                // owner of these entities is same
-                if (caller.getId() == Account.ACCOUNT_ID_SYSTEM || _accountMgr.isRootAdmin(caller.getId())) {
-                    if (ag.getAccountId() != owner.getAccountId()) {
-                        throw new PermissionDeniedException("Affinity Group " + ag + " does not belong to the VM's account");
+                if (ag.getAclType() == ACLType.Domain) {
+                    _accountMgr.checkAccess(caller, null, false, owner, ag);
+                    // Root admin has access to both VM and AG by default,
+                    // but
+                    // make sure the owner of these entities is same
```

Comment:
sameOwner flag is set to false in `_accountMgr.checkAccess()`. Update the above comment, as the owner of the entity is not the same.